### DISCUSSION TITLE
How to create a custom useQuery hook with optional configuration options (useQueryOption)❓

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [TanStack/query#3227] [Q&A] [answered]
Hi all,
I was trying to figure out a way to create a custom `useQuery` hook that could accept an optional object with `useQueryOption` values. This is useful for scenarious where you might use a hook in different places:

```ts
// Page1
const myData = useCustomHook('some_val', {
  onSuccess: () => navigate('/settings'),
  //....other valid options.....
})

// Page2
const myData = useCustomHook('some_val', {
  onSuccess: () => navigate('/other_route'),
    //....other valid options, but different.....
})
```

## What I've Tried 
Initially my custom hook using `useQuery` wasn't giving me the type inference I needed:

```ts
type Item = { name: string };
const fetcher = async (): Promise<Item[]> => [{ name: "adam" }, { name: "ben" }];

// has ability to pass optional values to useQuery's `useQueryOptions` object but  types are lost ❌
export const useGetData = (randomParam: string, configOptions: UseQueryOptions) => {
  const queryResponse = useQuery({
    queryKey: [randomParam],
    queryFn: fetcher,
    ...configOptions // ← this object spreading causing problems
  });
  return queryResponse; // ❌ unfortunately this comes back UseQueryResult<unknown, unknown> , it should be UseQueryResult<Item[], unknown>
};
```

I was able to get back the type information I needed by casting `useQuery`'s response like this:

```ts
type Item = { name: string };
const fetcher = async (): Promise<Item[]> => [{ name: "adam" }, { name: "ben" }];

export const useGetData = (randomParam: string, configOptions: UseQueryOptions) => {
  const queryResponse = useQuery({
    queryKey: [randomParam],
    queryFn: fetcher,
    ...configOptions
  }) as UseQueryResult<Item[]>;  // ✅
  return queryResponse;
};
```

Is there a way to get this to work without casting? I've tried this (no cast) but the issue seems to be the `...configOptions` being spread, which confuses typescript.

```ts
type Item = { name: string };
const fetcher = async (): Promise<Item[]> => [{ name: "adam" }, { name: "ben" }];

// --------------------------------------------------------------------------------------------↓
export const useGetData = (randomParam: string, configOptions: UseQueryOptions): UseQueryResult<Item[]>; => {
  const queryResponse = useQuery({
    queryKey: [randomParam],
    queryFn: fetcher,
    ...configOptions
  }) 
  return queryResponse;  // ❌ unfortunately this comes back UseQueryResult<unknown, unknown> , it should be UseQueryResult<Item[], unknown>
};
```

Codesandbox: https://codesandbox.io/s/nervous-engelbart-mzygc?file=/src/App.tsx

## References
- Thanks @TkDodo for this blog post:  https://tkdodo.eu/blog/react-query-and-type-script
- https://github.com/tannerlinsley/react-query/discussions/929

**Comment by mfpopa** [comment]
Type inference does not seem to work for `UseQueryOptions`, so you just need to specify a generic => `UseQueryOptions<Item[]>`.

```TypeScript
export const useGetData = (randomParam: string, configOptions: UseQueryOptions<Item[]>) => {
  return useQuery([randomParam], fetcher, {
    ...configOptions, // ← this object spreading causing problems
  });
};
```

Type inference is great when it works, but sometimes it does not and it can get very complicated which is why I prefer specifying types everywhere. I am not talking about React Query, but TypeScript code in general.

**Comment by TkDodo** [comment, marked as answer]
This is not a type inference issue. `UseQueryOptions` has 4 generics. If you just write `UseQueryOptions`, all default values for the generics are chosen:

https://github.com/TanStack/query/blob/250c654673953975c434692a0be7975a1b94b259/src/reactjs/types.ts#L26-L31

if that doesn't match your actual types, you get mismatches.

Passing one generic as @mfpopa suggests will solve most of the issues, but `select` won't work and the `queryKey` will not have the correct type in the `queryFunctionContext`.

My usual advice would be to _not_ create abstractions that allow passing all query options, but to only provide those that you want consumers to override. Otherwise, the abstraction is likely too wide, and you'd need to juggle a lot of generics like react-query has to do internally :)
Some props you likely don't want consumers of your custom hook to provide, like `cacheTime` or `staleTime`, as it doesn't make much sense to have different values for the same cache entry.

**Reply by mfpopa** [reply]
@TkDodo That is actually great advice. This also decouples the consumer components from React Query's types.

**Reply by hamanuha** [reply]
So would this be the correct way to only allow the `enable` option for a custom hook?
```ts
const useItems = (options?: Pick<UseQueryOptions<Item[], unknown, Item[], string[]>, 'enabled'>) => {
  return useQuery({
    queryKey: ['items'],
    queryFn: fetchItems,
    ...options,
  })
}
```

**Reply by NarenderSiri** [reply]
Please..tell me how to write in react js

**Reply by NarenderSiri** [reply]
How to write for useMutation

**Reply by NarenderSiri** [reply]
How to write in useMutation

**Reply by TkDodo** [reply]
> So would this be the correct way to only allow the enable option for a custom hook?

imo, you're overcomplicating things:

```
const useItems = (enabled?: boolean) => {
  return useQuery({
    queryKey: ['items'],
    queryFn: fetchItems,
    enabled
  })
}
```

or:

```
const useItems = (options?: { enabled?: boolean }) => {
  return useQuery({
    queryKey: ['items'],
    queryFn: fetchItems,
    ...options
  })
}
```

**Reply by thatCarlyleGuy** [reply]
@TkDodo If we're [encouraged](https://tkdodo.eu/blog/practical-react-query#create-custom-hooks) to create our own custom hooks, I think it's important that we're able to do that easily in Typescript. Specifically, that we're able to make use of type inference for the `select` field without having to pass explicit types to useQuery.

A possible way of doing this might look like this (using the v3 api in this case):
```typescript
const useQueryFactory =
  <TQueryFnData = unknown, TError = unknown, TData = TQueryFnData, TQueryKey extends QueryKey = QueryKey>(
    queryKey: TQueryKey,
    queryFn: QueryFunction<TQueryFnData, TQueryKey>,
  ) => <SelectType = TData>(
    options?: Omit<UseQueryOptions<TQueryFnData, TError, SelectType, TQueryKey>, 'queryKey' | 'queryFn'>
  ) => {
  return useQuery(
    queryKey,
    queryFn,
    options
  );
};

const useItemsQuery = useQueryFactory(
  ['items'],
  fetchItems,
);

const TypesInferred = () => {
  // type is inferred as UseQueryResult<Items[], unknown>
  const messageTemplates = useItemsQuery();

  // type is inferred as UseQueryResult<SomeItems[], unknown>
  const userTemplates = useItemsQuery({
    select: selectSomeItems
  });

  // ...
};
```

Would be really cool if something like this was available in the library given how [powerful](https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option) the select option is and how generalised we can make our custom hooks out of the gate.

**Reply by TkDodo** [reply]
> Specifically, that we're able to make use of type inference for the select field without having to pass explicit types to useQuery.

You don't need to do that. Here's how I write a custom hook that takes select:

```
type Items = Array<string> // whatever items is
const useItemsQuery = <TResult = Items>(select?: ((data: Items) => TResult)) =>
  useQuery({ queryKey, queryFn, select })

useItemsQuery()
useItemsQuery(items => items.length)
```

**Reply by thatCarlyleGuy** [reply]
That's what I'm trying to avoid. I'm maintaining a library with 30+ custom hooks. 
1. I want to standardise the use of the `select` option. Using your suggested method, I'd have to add the same generic type to each hook.
2. Given that my library is shared across the org, I don't know what the consumer of that hook will need down the line. I'd rather set sane defaults and expose the full list of options for each hook instead of having each dev tinker with the implementation each time they need an additional option for each hook. The `useQueryFactory` does that for free; unifies the interface exposed across all custom hooks and has full type inference.

**Reply by LeoCaprile** [reply]
> @TkDodo If we're [encouraged](https://tkdodo.eu/blog/practical-react-query#create-custom-hooks) to create our own custom hooks, I think it's important that we're able to do that easily in Typescript. Specifically, that we're able to make use of type inference for the `select` field without having to pass explicit types to useQuery.
> 
> A possible way of doing this might look like this (using the v3 api in this case):
> 
> ```ts
> const useQueryFactory =
>   <TQueryFnData = unknown, TError = unknown, TData = TQueryFnData, TQueryKey extends QueryKey = QueryKey>(
>     queryKey: TQueryKey,
>     queryFn: QueryFunction<TQueryFnData, TQueryKey>,
>   ) => <SelectType = TData>(
>     options?: Omit<UseQueryOptions<TQueryFnData, TError, SelectType, TQueryKey>, 'queryKey' | 'queryFn'>
>   ) => {
>   return useQuery(
>     queryKey,
>     queryFn,
>     options
>   );
> };
> 
> const useItemsQuery = useQueryFactory(
>   ['items'],
>   fetchItems,
> );
> 
> const TypesInferred = () => {
>   // type is inferred as UseQueryResult<Items[], unknown>
>   const messageTemplates = useItemsQuery();
> 
>   // type is inferred as UseQueryResult<SomeItems[], unknown>
>   const userTemplates = useItemsQuery({
>     select: selectSomeItems
>   });
> 
>   // ...
> };
> ```
> 
> Would be really cool if something like this was available in the library given how [powerful](https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option) the select option is and how generalised we can make our custom hooks out of the gate.

Man you're a life saver, I needed the options to keep the typing, I were debuggin for hours, and saw your code and a bulb turn on. Thanks!

**Reply by ArturKustyaev** [reply]
> @TkDodo If we're [encouraged](https://tkdodo.eu/blog/practical-react-query#create-custom-hooks) to create our own custom hooks, I think it's important that we're able to do that easily in Typescript. Specifically, that we're able to make use of type inference for the `select` field without having to pass explicit types to useQuery.
> 
> A possible way of doing this might look like this (using the v3 api in this case):
> ```typescript
> const useQueryFactory =
>   <TQueryFnData = unknown, TError = unknown, TData = TQueryFnData, TQueryKey extends QueryKey = QueryKey>(
>     queryKey: TQueryKey,
>     queryFn: QueryFunction<TQueryFnData, TQueryKey>,
>   ) => <SelectType = TData>(
>     options?: Omit<UseQueryOptions<TQueryFnData, TError, SelectType, TQueryKey>, 'queryKey' | 'queryFn'>
>   ) => {
>   return useQuery(
>     queryKey,
>     queryFn,
>     options
>   );
> };
> 
> const useItemsQuery = useQueryFactory(
>   ['items'],
>   fetchItems,
> );
> 
> const TypesInferred = () => {
>   // type is inferred as UseQueryResult<Items[], unknown>
>   const messageTemplates = useItemsQuery();
> 
>   // type is inferred as UseQueryResult<SomeItems[], unknown>
>   const userTemplates = useItemsQuery({
>     select: selectSomeItems
>   });
> 
>   // ...
> };
> ```
> 
> Would be really cool if something like this was available in the library given how [powerful](https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option) the select option is and how generalised we can make our custom hooks out of the gate.

Good afternoon, how can I pass query parameters to the queryFn function?

**Reply by WesleyYue** [reply]
Any idea on how to make a useInfiniteQueryFactory? Have been struggling on this for a while

**Comment by nextglabs** [comment]
Had a similar situation and I solved it this way by passing the same type used in `UseQueryOptions<T>` to the `useQuery<T>(...)`

```typescript
export const useGetData = (randomParam: string, configOptions?: UseQueryOptions<Item[]>) => {
  return useQuery<Item[]>([randomParam], fetcher, options);
};
```

**Reply by TkDodo** [reply]
This is a limited approach that will only get you so far - the `select` option won't work with that, for example. But great that it works for your case 👍

**Comment by devgogo** [comment]
resolved

```ts
export const useHealth = (options?: QueryOptions<HealthResult, ErrorResponse>) => {
  return useQuery(['health'], health, options)
}
```

and

```ts
export const useUnsubscribe = (
  options?: MutationOptions<unknown, ErrorResponse, string>
) => {
  return useMutation(unsubscribe, options)
}
```

**Comment by NarenderSiri** [comment]
Please …how to write in react js

**Comment by mkarajohn** [comment]
In case this is still an issue for anyone that stumbles upon this thread, I resolved this like so:

```typescript
export function useSpecificButCustomizableQuery(
  params: Record<string, string>,
  queryOptions: Omit<
    UseQueryOptions<
      TypeOfReturnValue,
      unknown,
      TypeOfReturnValue,
      TypeOfTheQueryKey
    >,
    'queryKey' | 'queryFn'
  > = {}
) {

  return useQuery<
    TypeOfReturnValue,
    unknown,
    TypeOfReturnValue,
    TypeOfTheQueryKey
  >({
    ...queryOptions,
    queryKey: ['your', 'key'],
    queryFn: (): Promise<TypeOfReturnValue> => {}
  });
}
```

**Comment by WillWarnerNerdery** [comment]
I utilized @TkDodo's approach with `Pick` to get some type safety.

```TS
 const useCustomQuery() (options: Pick<UseQueryOptions, 'enabled'> = { enabled: true }) => {
    useQuery( queryKey, queryFn, options)
}
```

**Comment by johnnyshankman** [comment]
minimum viable reproduction:
```ts
export const useQueryWithoutWindowRefetch = (options: UseQueryOptions) => {
  return useQuery({
    ...options,
   refetchOnWindowFocus: false,
};
```

you just can't really proxy this with premade preferences.

you lose the type inheritance since you dont the know the generics to pass to `UseQueryOptions`, it's always `unknown`

i dont think you can fix this in any simple fashion. may as well close and say not fixing.

**Reply by ArturKustyaev** [reply]
> minimum viable reproduction:
> 
> ```ts
> export const useQueryWithoutWindowRefetch = (options: UseQueryOptions) => {
>   return useQuery({
>     ...options,
>    refetchOnWindowFocus: false,
> };
> ```
> 
> you just cannot proxy this thing no matter what with premade preferences.
> 
> you lose the type inheritance bc you cannot possibly the know the generics to pass to `UseQueryOptions`, it's always `unknown`
> 
> i dont think you can fix this in any simple fashion. may as well close and say not fixing.

```ts
interface Data {
  id: number;
  title: string;
}

export const useQueryWithoutWindowRefetch = <TData = Data>(
  options?: Omit<UseQueryOptions<Data, DefaultError, TData>, 'queryKey' | 'queryFn'>,
) =>
  useQuery({
    queryKey: ['query keys'],
    queryFn: () => Promise.resolve({ id: 1, title: 'title' }),
    refetchOnWindowFocus: false,
    ...options,
  });

const { data } = useQueryWithoutWindowRefetch({ select: (data) => data.id });
```

**Reply by TkDodo** [reply]
as you can see, it is fixable. You "just" need to take a look at how react query has implemented type parameters internally and do the same thing. But it's still not advised. The best way to set premade preferences is `defaultOptions` (to target all queries) or with `queryClient.setQueryDefaults` (to target some queries matching a certain structure).

With v5, you can also use `queryOptions` to share configs around instead of writing custom hooks.

**Comment by ArturKustyaev** [comment]
Anyone who has problems with type inference, try my method. I slightly rewrote the QueryOptions and InfiniteQueryOptions types for convenience. The select function automatically infered types.

```ts
import {
  DefaultError,
  InfiniteData,
  QueryKey,
  useInfiniteQuery,
  UseInfiniteQueryOptions,
  UseMutationOptions,
  useQuery,
  UseQueryOptions,
} from '@tanstack/react-query';
import axios from 'axios';

export type QueryOptions<TQueryFnData = unknown, TData = TQueryFnData, TError = DefaultError> = Omit<
  UseQueryOptions<TQueryFnData, TError, TData>,
  'queryKey' | 'queryFn'
>;

export type InfiniteQueryOptions<
  TQueryFnData = unknown,
  TData = TQueryFnData,
  PageParam = number,
  TQueryData = TQueryFnData,
  TError = DefaultError,
> = Omit<
  UseInfiniteQueryOptions<TQueryFnData, TError, TData, TQueryData, QueryKey, PageParam>,
  'queryKey' | 'queryFn' | 'getNextPageParam' | 'initialPageParam'
>;

type Person = {
  id: number;
  name: string;
};

type QueryParams = {
  page?: number;
  count?: number;
};

type QueryResult = Awaited<ReturnType<typeof getPersons>>;

const getPersons = (params: QueryParams) => axios.get<Person[]>('get-persons', { params });

const useGetPersonsInfiniteQuery = <SelectData = InfiniteData<QueryResult>>(
  { page = 1, count = 30 }: QueryParams = {},
  options?: InfiniteQueryOptions<QueryResult, SelectData>,
) =>
  useInfiniteQuery({
    queryKey: ['query keys'],
    initialPageParam: page,
    queryFn: ({ pageParam }) => getPersons({ page: pageParam, count }),
    getNextPageParam: () => /* your get next param function */ undefined,
    ...options,
  });

const useGetPersonsQuery = <SelectData = QueryResult>(
  { page = 1, count = 30 }: QueryParams = {},
  options?: QueryOptions<QueryResult, SelectData>,
) =>
  useQuery({
    queryKey: ['query keys'],
    queryFn: () => getPersons({ page, count }),
    ...options,
  });

const { data } = useGetPersonsInfiniteQuery();
/*       ^ InfiniteData<AxiosResponse<Person[], any>, unknown> | undefined */

const { data } = useGetPersonsInfiniteQuery({}, { select: (data) => data.pages.flatMap((page) => page.data) });
/*       ^ data: Person[] | undefined */

const { data } = useGetPersonsQuery();
/*       ^ data: AxiosResponse<Person[], any> | undefined */

const { data } = useGetPersonsQuery({}, { select: (data) => data.data });
/*       ^ data: Person[] | undefined */


```
